backends/dpdk/dpdkProgram.cpp

```diff
@@ -52,6 +53,14 @@ const IR::DpdkAsmProgram *ConvertToDpdkProgram::create(IR::P4Program *prog) {
                     {new IR::Annotation(IR::ID("__metadata__"), {})});
                 for (auto anno : s->annotations->annotations)
                     annotations->add(anno);
+
+                /* Fetch temporary variable created earlier for holding b & c for &&& operation */
+                for (auto field : s->fields) {
+                   if ((field->name.name).find("tmpMask_b_AND_c")) {
+                       parserMask = new IR::PathExpression(IR::ID("m."+field->name.name));
```

Comment:
This isn't really safe, if the user happens to have a variable with this name you will use the wrong one.
Moreover, this is also wrong if you have multiple such variables with different types. Consider the case where you have two states with select statements with different types. Then you will have two such variables with different types.
Perhaps the simplest solution is to have a pass that converts a select into a series of cascading if statements.
The current IR cannot represent an if that transitions, so you would need to define a new IR node for that, but that's not particularly difficult.

Reply:
Yeah, I missed the point that searching for the variable using substring would be erroneous in case of multiple such variables and user defined variables.

And regarding writing a different pass for this: the select statement in dpdk is always converted into conditional jumps. I just made use of the existing framework to generate these conditional jumps to support Mask operator. Writing a new pass would be cleaner but would also require overall design change. 
@hanw any comments/concerns for this?

The issue with accessing variable name correctly could be fixed in current framework by defining a new map with "<blockname_statename> => path expression" as key=>value pair for these temp variables. And then, instead of using substring matching, access the variables using this map.

>> The current IR cannot represent an if that transitions, so you would need to define a new IR node for that, but that's not particularly difficult.

I don't quite understand this. Even in general case (without mask operation), the select transition statement is converted to Ifs, so not sure what difference it would make in IR for handling this case.

Reply:
1) I don't think your last proposal would work either. Consider the following:
`select (a, b) {  8w4 &&& 8w2, 16w4 &&& 16w 2: a }`
There are two masks in the same select.
2) You already have a translation for if, so if there is no select there is one fewer case to consider
3) Why do you even need to insert variable declarations in the P4 code? These variables are not really used anywhere. Can't you actually just insert dpdk variable declarations in your IR while you scan the select expression?

Reply:
Some existing compiler infra may help you on this:
-  https://github.com/p4lang/p4c/blob/main/ir/pattern.h help you to write code to match on a particular expression pattern, instead of hard-code string match on names.
-  findContext() to filter IR usage in a paritcular context, e.g. a mask expression under IR::TableKey.

Reply:
>> There are two masks in the same select.

Yes, I think the "key" used in my last proposal would have to be modified further (using the select expression probably) to uniquely identify each mask.

>> Why do you even need to insert variable declarations in the P4 code? These variables are not really used anywhere. 

The general assembly format for dpdk is:
\<structure declarations for header and metadata\>
\<instance of header and metadata\>
\<extern declarations\>
\<actions\>
\<tables\>
\<apply block\>

There is single apply block which consists of all the statements from parser and control blocks. All local variables from parser and control blocks are encapsulated within a single metadata struct and accessed using "m.<var_name>" within the apply block. So we need to insert the temporary variable declaration into the metadata struct.

>> Can't you actually just insert dpdk variable declarations in your IR while you scan the select expression?

Thanks for pointing that out. I had a deeper look in the pass where we scan the transition select expression. I do believe that I can modify the metadata struct in the same pass to insert the new temp variables.

So the question is: Should we update that pass for Mask handling or we absolutely require to pull this select expression handling out of current framework and write a new pass as you suggested earlier?

Reply:
It looks to me like you should be able to handle select masks in only one pass, without the preparation pass that inserts variables. Just insert the metadata dpdk information right when you need it, in the same pass that converts masks.

Reply:
I have modified the implementation to add the temp variables to metadata struct in the same pass where the select expressions are being scanned.
Some of the code is reordered in the  ConvertToDpdkProgram::create() function. This is to enable the new temp variables to be added to metadata before converting the metadata struct to DPDK IR.